src/protego.py

```diff
@@ -0,0 +1,406 @@
+
+
+class Protego(object):
+
+    def __init__(self):
+        # A dict mapping user agents (specified in robots.txt) to rule sets.
+        self._user_agents = {}
+
+        # Preferred host specified in the robots.txt
+        self._host = None
+
+        # A list of sitemaps specified in the robots.txt
+        self._sitemap_list = []
+
+        # A memoization table mapping user agents (used in queries) to matched rule sets.
+        self._matched_rule_set = {}
+
+        self._total_line_seen = 0
+        self._invalid_directive_seen = 0
+        self._total_directive_seen = 0
+
+    @classmethod
+    def parse(cls, content):
+        o = cls()
+        o._parse_robotstxt(content)
+        return o
+
+    def _parse_robotstxt(self, content):
+        lines = content.splitlines()
+
+        # A list containing all the user agents of the current record group.
+        current_user_agents = []
+
+        # Last encountered rule irrespective of whether it was valid or not.
+        previous_rule_field = None
+
+        for line in lines:
+            self._total_line_seen += 1
+
+            # Remove the comment portion of the line
+            hash_pos = line.find('#')
+            if hash_pos != -1:
+                line = line[0: hash_pos].strip()
+
+            # Whitespace at the beginning and at the end of the line is ignored.
+            line = line.strip()
+            if not line:
+                continue
+
+            # Format for a valid robots.txt rule is "<field>:<value>"
+            if line.find(':') != -1:
+                field, value = line.split(':', 1)
+            else:
+                # We will be generous here and give it a second chance.
+                parts = line.rsplit(' ', 1)
+                if not len(parts) == 2:
+                    continue
+                field, value = parts
+
+            field = field.strip().lower()
+            value = value.strip()
+
+            # Ignore rules with no value part (e.g. "Disallow: ", "Allow: ").
+            if not value:
+                previous_rule_field = field
+                continue
+
+            # Ignore rules without a corresponding user agent.
+            if not current_user_agents and field not in _USER_AGENT_DIRECTIVE:
+                logger.debug("Rule at line {} without any user agent to enforce it on.".format(self._total_line_seen))
+                continue
+
+            self._total_directive_seen += 1
+
+            if field in _USER_AGENT_DIRECTIVE:
+                if previous_rule_field and previous_rule_field not in _USER_AGENT_DIRECTIVE:
+                    current_user_agents = []
+
+                # Wildcards are not supported in the user agent values.
+                # We will be generous here and remove all the wildcards.
+                user_agent = value.strip().lower()
+                if user_agent != '*' and '*' in user_agent:
+                    user_agent = user_agent.replace('*', '')
```

Comment:
In this case, shouldn‘t we consider the user agent value both with and without the asterisk?

I mean, if you get `User-Agent: asdf*asdf`, the person who wrote the `robots.txt` file may have meant `*` either as a literal asterisk or as a wildcard. I get that we don‘t support the asterisk as a wildcard in the user agent string. But, if I’m reading this code right, here we don’t support the other potential user intention either. As we are turning `asdf*asdf` into `asdfasdf`, a user agent string that is literally `asdf*asdf` will not match this ruleset.